File: 1642. Furthest Building You Can Reach.py
# 1642. Furthest Building You Can Reach

class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        hp = [] # max heap

        for i in range(len(heights)-1):
            gap = heights[i+1] - heights[i]
            if gap <= 0:
                continue
            bricks -= gap
            heappush(hp, -gap)

            if bricks < 0:
                if not ladders:
                    return i
                else:
                    ladders -= 1
                    if hp:
                        bricks -= heappop(hp)
            
        return len(heights) - 1

File: 1642. Furthest Building You Can Reach.cpp
// 1642. Furthest Building You Can Reach

#include <vector>
#include <queue>
using namespace std;

class Solution {
    public:
        int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
            priority_queue<int, vector<int>> maxHeap; // Min-heap to store ladder jumps
            int n = heights.size();
            
            for (int i = 0; i < n - 1; i++) {
                int diff = heights[i + 1] - heights[i];
                if (diff > 0) {
                    bricks -= diff;
                    maxHeap.push(diff);
                    if (bricks < 0) {
                        if (ladders > 0) {
                            bricks += maxHeap.top();
                            maxHeap.pop();
                            ladders -= 1;
                        }
                        else return i; // Stop if out of bricks
                    }
                }
            }
            return n - 1; // Reached the last building
        }
    };